marlowe/src/Language/Marlowe/Semantics.hs

```diff
@@ -107,108 +156,268 @@ data Observation = AndObs Observation Observation
+                   | ReducePartialPay AccountId Payee Money Money
+                                   -- ^ src    ^ dest ^ paid ^ expected
+                   | ReduceShadowing ValueId Integer Integer
+                                    -- oldVal ^  newVal ^
+  deriving (Show)
+
+
+-- | Result of 'reduceContractStep'
+data ReduceStepResult = Reduced ReduceWarning ReduceEffect State Contract
+                      | NotReduced
+                      | AmbiguousSlotIntervalReductionError
+  deriving (Show)
+
+
+-- | Result of 'reduceContractUntilQuiescent'
+data ReduceResult = ContractQuiescent [ReduceWarning] [Payment] State Contract
+                  | RRAmbiguousSlotIntervalError
+  deriving (Show)
+
+
+-- | Warning of 'applyCases'
+data ApplyWarning = ApplyNoWarning
+                  | ApplyNonPositiveDeposit Party AccountId Integer
+  deriving (Show)
+
+
+-- | Result of 'applyCases'
+data ApplyResult = Applied ApplyWarning State Contract
+                 | ApplyNoMatchError
+  deriving (Show)
+
+
+-- | Result of 'applyAllInputs'
+data ApplyAllResult = ApplyAllSuccess [TransactionWarning] [Payment] State Contract
+                    | ApplyAllNoMatchError
+                    | ApplyAllAmbiguousSlotIntervalError
+  deriving (Show)
+
+
+-- | Warnings during transaction computation
+data TransactionWarning = TransactionNonPositiveDeposit Party AccountId Integer
+                        | TransactionNonPositivePay AccountId Payee Integer
+                        | TransactionPartialPay AccountId Payee Money Money
+                                               -- ^ src    ^ dest ^ paid ^ expected
+                        | TransactionShadowing ValueId Integer Integer
+                                                -- oldVal ^  newVal ^
+  deriving (Show)
+
+
+-- | Transaction error
+data TransactionError = TEAmbiguousSlotIntervalError
+                      | TEApplyNoMatchError
+                      | TEIntervalError IntervalError
+                      | TEUselessTransaction
+  deriving (Show)
+
+
+{-| Marlowe transaction input.
+-}
+data TransactionInput = TransactionInput
+    { txInterval :: SlotInterval
+    , txInputs   :: [Input] }
+  deriving (Show)
+
+
+{-| Marlowe transaction output.
+-}
+data TransactionOutput =
+    TransactionOutput
+        { txOutWarnings :: [TransactionWarning]
+        , txOutPayments :: [Payment]
+        , txOutState    :: State
+        , txOutContract :: Contract }
+    | Error TransactionError
+  deriving (Show)
 
 
+{-|
+    This data type is a content of a contract's /Data Script/
+-}
+data MarloweData = MarloweData {
+        marloweCreator  :: Party,
+        marloweState    :: State,
+        marloweContract :: Contract
+    } deriving (Show)
+
+
+-- | Empty State for a given minimal 'Slot'
+emptyState :: Slot -> State
+emptyState sn = State
+    { accounts = Map.empty ()
+    , choices  = Map.empty ()
+    , boundValues = Map.empty ()
+    , minSlot = sn }
+
+-- | Returns an owner of an account
```

Comment:
I was about to write "why don't you use records for these types, then?" and then I remembered that they're effectively the "Marlowe syntax". I think that's worth a note.